qiskit_experiments/library/randomized_benchmarking/mirror_rb_experiment.py

```diff
@@ -0,0 +1,522 @@
+                                    the backend coupling map.
+            backend: The backend to run the experiment on.
+            num_samples: Number of samples to generate for each
+                         sequence length.
+            seed: Optional, seed used to initialize ``numpy.random.default_rng``.
+                  when generating circuits. The ``default_rng`` will be initialized
+                  with this seed value everytime :meth:`circuits` is called.
+            full_sampling: If True all Cliffords are independently sampled for
+                           all lengths. If False for sample of lengths longer
+                           sequences are constructed by appending additional
+                           Clifford samples to shorter sequences.
+            inverting_pauli_layer: If True, a layer of Pauli gates is appended at the
+                                   end of the circuit to set all qubits to 0 (with
+                                   possibly a global phase)
+
+        Raises:
+            QiskitError: if an odd length or a negative two qubit gate density is provided
+        """
+        # All lengths must be even
+        if not all(length % 2 == 0 for length in lengths):
+            raise QiskitError("All lengths must be even")
+
+        # Two-qubit density must be non-negative
+        if two_qubit_gate_density < 0:
+            raise QiskitError("Two-qubit gate density must be non-negative")
+
+        super().__init__(
+            qubits,
+            lengths,
+            backend=backend,
+            num_samples=num_samples,
+            seed=seed,
+            full_sampling=full_sampling,
+        )
+
+        self._local_clifford = local_clifford
+        self._pauli_randomize = pauli_randomize
+        self._two_qubit_gate_density = two_qubit_gate_density
+
+        # By default, the inverting Pauli layer at the end of the circuit is not added
+        self._inverting_pauli_layer = inverting_pauli_layer
+
+        # Set analysis options
+        self.analysis = MirrorRBAnalysis()
+
+    def _sample_circuits(self, lengths, rng) -> List[QuantumCircuit]:
+        """Sample Mirror RB circuits.
+
+        Steps:
+        1. Sample length/2 layers of random Cliffords
+        2. Compute inverse of each layer in the first half of the circuit and append to circuit
+        3. Sample the random paulis and interleave them between the Clifford layers
+        4. Sample the 1-qubit local Clifford and add them to the beginning and end of the circuit
+
+        Args:
+            lengths: List of lengths to run Mirror RB
+            rng: Generator seed
+
+        Returns:
+            List of QuantumCircuits
+
+        Raises:
+            QiskitError: if backend without a coupling map is provided
+        """
+
+        # Backend must have a coupling map
+        if not self._backend or not self._backend.configuration().coupling_map:
```

Comment:
Hmm... I understand why we need this check here. Basically MirrorRB can target more than 2Q, then we should avoid applying 2Q gate to non-adjacent qubits otherwise extra SWAP is required. 

On the other hand, the standard execution model of QE for now is
1. define circuit in the virtual qubit domain
2. call transpiler with initial layout, which is based on the experiment's physical qubits
3. run circuit in the physical qubit domain (transpiled circuit) on the target backend. 

and this experiment intentionally avoids step 1 due to above constraint. This probably motivate us to turn the `backend` in the experiment class constructor into required (currently it's optional), and to deprecate another `backend` argument in the `BaseExperiment.run` method. This model indicates an experiment instance is always aware of the target backend, and we can reasonably generate circuits on top of some hardware architecture constraints -- as you are doing in this PR.

Reply:
I see, the reason why we added this check instead of making the backend required is because @chriseclectic and @coruscating had mentioned that backends aren't serializable, so it may be better to keep it optional and only check for it when the circuits are being generated. Considering these points, what would be the best option here?

Reply:
I lean toward having backend in the constructor as required because this benefits other experiments as well (e.g. T1 experiment needs to check timing constraints to build circuits). I think we should add special handling for backend in serialization/deserialization (I think it's enough to keep backend name instead of serialized object and ask user to set identical backend if the experiment is loaded instance).

Reply:
We discussed several potential plans for handling this in the meeting. @nkanazawa1989 are you okay with keeping the current implementation for this experiment and merging first, and then working on changing the implementation in a future PR?